### PR TITLE
Improve KE for commandline and programmatically tuning dispatch

### PR DESCRIPTION
**Pull request by cloudhan**

### CODE REVIEW

**Comment by cloudhan** [comment]
For example:

```python
import os
import sys

sys.path.insert(0, "/home/guangyunhan/onnxruntime/onnxruntime/python/tools/kernel_explorer/kernels")
sys.path.insert(0, "/home/guangyunhan/onnxruntime/build_rocm/Release")
os.environ["KERNEL_EXPLORER_BUILD_DIR"] = "/home/guangyunhan/onnxruntime/build_rocm/Release"


import multiprocessing as mp
from multiprocessing import Pool, current_process


def profile(name, *args, **kwargs):
    import kernel_explorer as ke

    ke.set_return_tuning_results()
    ke.set_dispatchable_pattern("*Tunable*")
    print(os.environ["HIP_VISIBLE_DEVICES"])
    if name == "gemm":
        from gemm_test import profile_with_args as profile

        return profile(*args, **kwargs)
    elif name == "softmax":
        from softmax_test import profile_with_args as profile

        return profile(*args, **kwargs)
    else:
        return []


def init():
    pidx = int(current_process()._identity[0]) - 1
    start_gpu = 2
    num_gpu = 14
    os.environ["HIP_VISIBLE_DEVICES"] = str(pidx % num_gpu + start_gpu)


if __name__ == "__main__":
    configs = [
        ("gemm", "float16", False, False, 1, 8912, 8912),
        ("gemm", "float16", False, False, 8, 8912, 8912),
        ("gemm", "float16", False, False, 16, 8912, 8912),
        ("gemm", "float16", False, False, 24, 8912, 8912),
        ("gemm", "float16", False, False, 32, 8912, 8912),
        ("gemm", "float16", False, False, 40, 8912, 8912),
        ("gemm", "float16", False, False, 48, 8912, 8912),
        ("softmax", 1, 1024, False, "float16"),
        ("softmax", 2, 1024, False, "float16"),
    ]

    mp.set_start_method("spawn")

    with Pool(processes=4, initializer=init) as pool:
        ret = pool.starmap(profile, configs, chunksize=1)

    from pprint import pprint
    from onnxruntime.tools.offline_tuning import Merger

    m = Merger()
    for tr in ret:
        m.merge(tr)

    pprint(m.get_merged())
```

File: onnxruntime/python/tools/kernel_explorer/kernels/kernel_explorer.py [hunk 1]
```diff
@@ -146,3 +182,182 @@ def benchmark(sort):
...
+                if ret is not None:
+                    print(
+                        f"WARNING: kernel explorer wants to override the return value of {f.__name__},",
+                        "but original return value is not None!",
+                    )
+                    return ret
+                try:
+                    trs = _kernel_explorer.get_collected_tuning_results()
+                finally:
+                    return trs
```
Comment (mindest) [inline onnxruntime/python/tools/kernel_explorer/kernels/kernel_explorer.py:323]: Do we want to catch possible exceptions here? In that case `trs` can be undefined.

File: onnxruntime/python/tools/kernel_explorer/kernels/gemm_test.py [hunk 2]
```diff
@@ -11,10 +11,12 @@
 import numpy as np
 import pytest
 from utils import dtype_to_suffix, get_gemm_basic_sizes, get_gemm_bert_sizes, get_gemm_bound, matmul, transab_to_suffix
+from ml_dtypes import float8_e4m3fn
```
Comment (mindest) [inline onnxruntime/python/tools/kernel_explorer/kernels/gemm_test.py:14]: currently unused?

**Review by mindest — COMMENTED** [review]
Thanks for this PR. Tested locally, it worked fine. Please fix the lint warnings/errors.

**Review by kailums — DISMISSED** [review]

**Review by mindest — APPROVED** [review]
Thanks for the new feature support.

**Review by mindest — DISMISSED** [review]

**Review by mindest — APPROVED** [review]

**Pull request merged**